New integration for Kostal Plenticore solar inverters

stegm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds an integration for Kostal Plenticore solar inverters (https://www.kostal-solar-electric.com/de-de/produkte/solar-wechselrichter/plenticore-plus). This component reads process data and setting values via REST-API and is also capable to change some of the inverter settings (battery minimal SoC for example). It uses a library which was inspired by an app https://github.com/kilianknoll, but the one for this integration uses async requests and has a dynamic data model which allows it to use it with different firmware versions.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

This component uses Config Flow for configuration. Two information are needed: Host or IP and password.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15764

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

stegm committed:
New integration for Kostal Plenticore solar inverters.

homeassistant commented:
Hi @stegm, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#43404).

Once you do that we will be able to review and accept this pull request.

Thanks!

stegm committed:
Fix errors from github pipeline.

stegm committed:
Fixed test for py37.

stegm committed:
Add more test for test coverage check.

stegm committed:
Try to fix test coverage check.

stegm committed:
Fix import sort order.

stegm committed:
Try fix test code coverage .

stegm committed:
Mock api client for tests.

pergolafabio commented:
this will not work for kostal piko covertors, right?

stegm commented:
> this will not work for kostal piko covertors, right?

@pergolafabio I'm not sure, but as far as I know from the https://www.photovoltaikforum.com the piko inverters uses a different API. Perhaps you can try https://www.msxfaq.de/sonst/iot/kostal15.htm.

pergolafabio commented:
i am using this custom  right now, not sure if its using the same api, or just webscraping
https://github.com/gieljnssns/kostalpiko-sensor-homeassistant

stegm commented:
> i am using this custom right now, not sure if its using the same api, or just webscraping
> https://github.com/gieljnssns/kostalpiko-sensor-homeassistant

The KostalPyko library which is used in their integration uses a different method as the one for the plenticore.

pergolafabio commented:
Make you can integrate it also? That's  makes a new integration possible for older invertor too :-)

No need for custom then

stegm commented:
> Make you can integrate it also? That's makes a new integration possible for older invertor too :-)

I think devices with separate communication libraries should be implemented in separate integrations. I have no possiblity to test other devices. Besides there will be only little common code between the two integrations.

pergolafabio commented:
Ok, np :-)

stegm committed:
Fix typo.

stegm committed:
Fix order of rebased code from dev.

stegm committed:
Add new data point for home power.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/__init__.py:
```diff
@@ -0,0 +1,287 @@
+            await self._client.logout()
+            _LOGGER.info("Logged out from %s.", self._config[CONF_HOST])
+
+    def register_entity(self, entity: PlenticoreProcessDataSensor) -> None:
```
This approach will need to be redesigned. The entity is a platform concern. We're not allowed to expose the entity API outside the entity. Only the core entity platform helper is allowed to access entities from outside.

Probably separate the device api and the entity api. If we need to notify an entity to turn unavailable from outside, we can use our dispatch helper.

But I don't see why we're doing it this way in the first place. The coordinator helper is built to fetch all relevant data and then notify connected listeners (entities) to grab their data from the data cache. We should follow this design.

I would create one coordinator per scope. The coordinator will automatically keep track of connected listeners (entities) and stop the update polling if there are no connected listeners.

The entities can check themselves if their data is available in the fetched data cache and modify their availability accordingly.

stegm replied:
I am not sure I have understood all the concerns.

I definitely agree with creating one coordinator per scope, so no problem with that.

But I don't understand exactly what you mean by "expose the entity API". Can you be a bit more specific? I am also not sure what you mean by "separate the device api and the entity api".

What I basically want to achieve is this: The coordinator should only query data points that have been added to the `hass` object (i.e. no deactivated entities) and secondly, data points that are not available due to the firmware version should not be queried (`available` property).

Feel free to show me an integration that implements this correctly in a similar way as a blueprint for this integration.

MartinHjelmare replied:
Exposing the entity API is allowing something else than the entity to access any of its attributes directly.

The default behavior of the coordinator is to only refresh the data if there are listeners attached. So disabling entities will work as expected.

If we know that an entity should never update due to missing firmware we can either make sure to not add it from the beginning or let it remove itself if it has been added. There's no point in adding an entity that will never be operational.

stegm replied:
- Use separate coordinator for each scope
- do not expose entity API to coordinator

MartinHjelmare commented on homeassistant/components/kostal_plenticore/sensor.py:
```diff
@@ -0,0 +1,294 @@
+        """Return the scope of this Sensor Entity."""
+        return SCOPE_SETTING
+
+    async def async_set_new_value(self, value) -> None:
```
Sensor entities should normally be read only.

stegm replied:
The entity is actually read-only. The method is called by the service to send a new value to the inverter via the API. Of course, this ultimately leads to the value of the entity also being changed.
I have adopted this pattern from other integrations, e.g. from Sonos. Please give me more information on how this can be implemented correctly.

stegm replied:
Removed write access via service from this PR for now.

MartinHjelmare reviewed (changes requested):
First impression is that the integration looks complicated and not in line with our design standards. I would advice to try and simplify this, at least for the first PR. Remove the writing of data and just focus on getting the reading (data fetching) done good.

stegm commented:
> Remove the writing of data and just focus on getting the reading (data fetching) done good.

I have no problem removing the writing of data for the time being. But on the other hand, this function has very little additional code and on the other hand, this was a main reason to do this integration in the first place, because it gives you the possibility to control the battery usage.

So if there is a way to integrate this I would be very happy.

stegm committed:
Modifications to review.
Remove service for write access (for first pull request). Refactor
update coordinator to not use the entity API.

stegm committed:
Fixed mock imports.

stegm committed:
Ignore new python module on coverage.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/config_flow.py:
```diff
@@ -0,0 +1,81 @@
+                try:
+                    await test_connection(self.hass, user_input)
+
+                    return self.async_create_entry(
```
Please only wrap the line that can raise in the try... except block. Use an `else:` block if needed. Here we can probably check the errors dict though to know if we should create the entry.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/config_flow.py:
```diff
@@ -0,0 +1,81 @@
+                    )
+                except PlenticoreAuthenticationException as ex:
+                    self._errors[CONF_PASSWORD] = "invalid_auth"
+                    _LOGGER.exception("Error response: %s", ex.msg)
```
Only non expected errors should be logged with stack trace. Use error level otherwise.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/config_flow.py:
```diff
@@ -0,0 +1,81 @@
+        self._errors = {}
+
+        if user_input is not None:
+            if user_input[CONF_HOST] not in configured_instances(self.hass):
```
If we invert this check and abort if true, we can save indentation.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/config_flow.py:
```diff
@@ -0,0 +1,81 @@
+                    await test_connection(self.hass, user_input)
+
+                    return self.async_create_entry(
+                        title=user_input[CONF_NAME], data=user_input
```
Use the host as entry title.

stegm replied:
I changed this to use the read hostname of the inverter, because if I use the host and the host is an IP the entity gets strange looking names like `sensor.192.168.1.1_home_power`.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/const.py:
```diff
@@ -0,0 +1,545 @@
+ATTR_VALUE = "value"
+ATTR_ENABLED_DEFAULT = "entity_registry_enabled_default"
+
+"""Defines all entities for process data.
```
Please write comments with `# `. Multi-line strings are not used for code comments.

MartinHjelmare commented on tests/components/kostal_plenticore/test_config_flow.py:
```diff
@@ -0,0 +1,165 @@
+    )
+
+    with patch(
+        "homeassistant.components.kostal_plenticore.config_flow.configured_instances",
```
Please add a MockConfigEntry instead.

MartinHjelmare reviewed (commented):
Nice redesign!

stegm committed:
Changes after review.

stegm commented on homeassistant/components/kostal_plenticore/__init__.py:
```diff
@@ -0,0 +1,70 @@
+    def shutdown(event):
+        hass.async_create_task(plenticore.logout())
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, shutdown)
```
I moved the shutdown logic to the `helper.Plenticore` class. Makes it easier to store the unsubscribe method.

stegm committed:
Fixed unit test because of config title.

stegm commented:
@MartinHjelmare Thank you for the  review! I learned a lot.

menkej commented:
What is the status of this PR, @MartinHjelmare? It seems that all requested changes have been done. I'd really love to integrate my Plenticore into HA...

MartinHjelmare commented:
Sorry, I'll get back to this asap.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -0,0 +1,276 @@
+    async def _fetch_data(self) -> Dict[str, Dict[str, str]]:
+        client = self._plenticore.client
+
+        if len(self._fetch) == 0 or client is None:
```
```suggestion
        if not self._fetch or client is None:
```

MartinHjelmare commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -0,0 +1,276 @@
+        self.hass = hass
+
+        # cache for the fetched data
+        self._data = {}
```
We don't need this dict. We fetch new data and return it every update.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -0,0 +1,276 @@
+        except (TypeError, ValueError):
+            return state
+
+        return PlenticoreDataFormatter.EM_STATES.get(value, "Unknown")
```
We use `None` to represent unknown state.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/sensor.py:
```diff
@@ -0,0 +1,192 @@
+    async_add_entities(entities)
+
+
+class PlenticoreDataSensor(CoordinatorEntity):
```
Also inherit from `SensorEntity` imported from the sensor integration.
```suggestion
class PlenticoreDataSensor(CoordinatorEntity, SensorEntity):
```

MartinHjelmare commented on tests/components/kostal_plenticore/test_config_flow.py:
```diff
@@ -0,0 +1,170 @@
+    )
+
+    with patch(
+        "homeassistant.components.kostal_plenticore.config_flow.test_connection",
```
Patch the client library instead of the config flow code.

MartinHjelmare commented on tests/components/kostal_plenticore/test_config_flow.py:
```diff
@@ -0,0 +1,170 @@
+    )
+
+    with patch(
+        "homeassistant.components.kostal_plenticore.config_flow.test_connection",
```
As above.

MartinHjelmare commented on tests/components/kostal_plenticore/test_config_flow.py:
```diff
@@ -0,0 +1,170 @@
+    )
+
+    with patch(
+        "homeassistant.components.kostal_plenticore.config_flow.test_connection",
```
As above.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/const.py:
```diff
@@ -0,0 +1,542 @@
+    (
+        "devices:local",
+        "Battery:DynamicSoc:Enable",
+        "Battery Dynamic Soc",
```
What kind of measurement is this?

stegm replied:
The Plenticore supports a battery connected to the DC3 string. There seems to be a function which manages the SOC dynamically depending on some previous state. This function can be enabled or disabled and this state can be read with this sensor.

MartinHjelmare replied:
If it's a binary sensor it should be part of a binary sensor platform.

MartinHjelmare commented on homeassistant/components/kostal_plenticore/sensor.py:
```diff
@@ -0,0 +1,192 @@
+    @property
+    def unique_id(self) -> str:
+        """Return the unique id of this Sensor Entity."""
+        return f"{self.entry_id}_{self._sensor_name}"
```
I suggest we use `self.data_id` as suffix. It's a bit safer. If we ever want to change the name of a sensor we don't need to change the unique id.

stegm replied:
good point!

stegm committed:
Fixes from review.

stegm commented on homeassistant/components/kostal_plenticore/helper.py:
```diff
@@ -0,0 +1,276 @@
+        if self._shutdown_remove_listener:
+            self._shutdown_remove_listener()
+
+        if self._login:
```
`login` is set to true only if the login was successful and hence logout has only to be executed in this case.

MartinHjelmare replied:
Right, but we'll never unload unless we've been able to login first.

stegm committed:
Changes from review (unique id and mocking of tests)

MartinHjelmare commented:
The branch is tainted with unrelated commits now. Please rebase and remove those commits.

stegm committed:
Use async update method. Change unique id. Remove _dict

stegm committed:
Remove _data field.

stegm committed:
Removed login flag from PlenticoreUpdateCoordinator.

stegm committed:
Removed Dynamic SoC sensor because it should be a binary sensor.

stegm committed:
Remove more sensors because they are binary sensors.

stegm commented on homeassistant/components/kostal_plenticore/const.py:
```diff
@@ -0,0 +1,535 @@
+    ),
+    (
+        "devices:local",
+        "EM_State",
```
This indicates the state of the energy manager which can have six different states.

MartinHjelmare reviewed (approved):
Great!
